packages/core/src/tools/injector.ts

```diff
@@ -0,0 +1,75 @@
+export interface Component<Instance, Deps extends any[]> {
+  (...args: Deps): Instance
+  $deps: Readonly<{ [K in keyof Deps]: AnyComponent<Deps[K]> }>
+}
+
+interface SimpleComponent<Instance> {
+  (): Instance
+}
+
+export type AnyComponent<Instance> = Component<Instance, any> | SimpleComponent<Instance>
+
+export interface Injector {
+  run: <T>(component: AnyComponent<T>) => T
+  get: <T>(component: AnyComponent<T>) => T
+  override: <T>(originalComponent: AnyComponent<T>, newComponent: AnyComponent<T>) => void
+  stop: () => void
+}
+
+export function createInjector(): Injector {
+  const instances = new Map<AnyComponent<any>, any>()
+  const overrides = new Map<AnyComponent<any>, AnyComponent<any>>()
+  overrides.set(getInjector, () => injector)
+
+  const injector: Injector = {
+    run(component) {
+      if (instances.has(component)) {
+        throw new Error(`Component ${component.name} already started`)
```

Comment:
Note: This assertion and the following ones (in `get` and `override`) are not strictly needed, but helps to debug weird injection cases.